File: S2P2_herencia/P2A4Empleado/P2A4Python/TrabajadorPiezas.py
from Empleado import Empleado

class TrabajadorPiezas(Empleado):
    def __init__(self, nombre, apellido, nss, salarioPorPieza, piezas):
        super().__init__(nombre, apellido, nss)
        self._salarioPorPieza = 0.0
        self._piezas = 0
        self.establecerSalarioPorPieza(salarioPorPieza)
        self.establecerPiezas(piezas)

    def establecerSalarioPorPieza(self, salario):
        self._salarioPorPieza = 0.0 if (salario < 0.0) else salario

    def obtenerSalarioPorPieza(self):
        return self._salarioPorPieza

    def establecerPiezas(self, piezasProducidas):
        self._piezas = 0 if (piezasProducidas < 0) else piezasProducidas

    def obtenerPiezas(self):
        return self._piezas

    def Ingresos(self):
        return self.obtenerSalarioPorPieza() * self.obtenerPiezas()

    def __str__(self):
        return f"trabajador por piezas: {super().__str__()}\nsalario por pieza: ${self.obtenerSalarioPorPieza():,.2f}; piezas producidas: {self.obtenerPiezas()}"

File: S2P2_herencia/P2A4Empleado/P2A4Java/TrabajadorPiezas.java
// Nueva clase: TrabajadorPiezas
// Representa a un empleado que cobra por pieza producida

public class TrabajadorPiezas extends Empleado {
    private double salarioPorPieza; // pago por cada pieza
    private int piezas; // número de piezas producidas

    // constructor
    public TrabajadorPiezas(String nombre, String apellido, String nss,
            double salarioPorPieza, int piezas) {
        super(nombre, apellido, nss);
        establecerSalarioPorPieza(salarioPorPieza);
        establecerPiezas(piezas);
    }

    // establecer salario por pieza
    public void establecerSalarioPorPieza(double salario) {
        salarioPorPieza = (salario < 0.0) ? 0.0 : salario;
    }

    // obtener salario por pieza
    public double obtenerSalarioPorPieza() {
        return salarioPorPieza;
    }

    // establecer piezas producidas
    public void establecerPiezas(int piezasProducidas) {
        piezas = (piezasProducidas < 0) ? 0 : piezasProducidas;
    }

    // obtener piezas producidas
    public int obtenerPiezas() {
        return piezas;
    }

    // implementar método abstracto ingresos
    @Override
    public double ingresos() {
        return salarioPorPieza * piezas;
    }

    // toString
    @Override
    public String toString() {
        return String.format(
                "trabajador por piezas: %s\n%s: $%,.2f; %s: %d",
                super.toString(),
                "salario por pieza", obtenerSalarioPorPieza(),
                "piezas producidas", obtenerPiezas());
    }
}